src/object/srv_ec_aggregate.c

```diff
@@ -256,13 +256,13 @@ agg_clear_extents(struct ec_agg_entry *entry)
 		}
 
 		if (extent->ae_orig_recx.rx_idx + extent->ae_orig_recx.rx_nr >
-		    next_stripe_st) {
+		    next_stripe_st && !tail) {
```

Comment:
I don't quite understand the tail and holdover extent logic here. could you please explain why need to add "!tail" here?  at above L248 ~ L255 you only init/set the extent when "if (tail)", why here (L259) add "!tail" then?

and seems I completely did not understand why need to remove those holdover extents?
thanks

Reply:
If tail is true, the visible portion of the extent carries over to the first of the next stripe.  If tail is false, the visible portion doesn't carry over but the orig extent goes beyond the end of the stripe.  Adding these to holdover extents allows these to be retained for subsequent stripes while enabling their removal once they're no longer needed.

The problem is that EC aggregation deals with visible extents, but the VOS extent remove has to be invoked at the granularity of the original extent.

I fixed this in this patch, because I caught the error, but it has nothing to do with range punch handling in EC aggregation.

Reply:
If the tail is zero, and orig_recx.rx_idx + orig_recx.rx_nr > start-of-next-stripe, it means the visible extents ends within this stripe by the orig extent goes into the next stripe.

Extent remove has to be applied to the entire orig extent.  If the visible portion spans the stripe boundary, the start offset is moved to the first of the next stripe, and extent is retained in the extent list.

Otherwise, if the orig extent is retained as a holdover extent.  If it's visible further into the next stripe, a subsequent visible extent (with the same orig extent specs) will be returned by the iterator.  This code ensures the (orig) extent will be removed in either case (i.e., visible in the next stripe or not).